source/ZipFile.cpp

```diff
@@ -0,0 +1,149 @@
+/* ZipFile.cpp
+Copyright (c) 2025 by tibetiroka
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "ZipFile.h"
+
+#include "Files.h"
+
+#include <functional>
+#include <numeric>
+
+using namespace std;
+
+
+ZipFile::ZipFile(const filesystem::path &zipPath) : basePath(zipPath), topLevelDirectory(), lock()
+{
+	lock_guard<recursive_mutex> guard(lock);
+	zipFile = unzOpen(basePath.string().c_str());
+	if(!zipFile)
+		throw runtime_error("Failed to open ZIP file" + zipPath.generic_string());
+
+	// Check whether this zip has a single top-level directory (such as high-dpi.zip/high-dpi)
+	filesystem::path topLevel;
+	for(const filesystem::path &path : ListFiles("", true, false))
+	{
+		filesystem::path zipPath = path.lexically_relative(basePath);
+		if(topLevel.empty())
+			topLevel = *zipPath.begin();
+		else if(*zipPath.begin() != topLevel)
+			return;
+	}
+	topLevelDirectory = topLevel;
+}
+
+
+
+ZipFile::~ZipFile()
+{
+	if(zipFile)
+	{
+		lock_guard<recursive_mutex> guard(lock);
+		unzClose(zipFile);
+	}
+}
+
+
+
+vector<filesystem::path> ZipFile::ListFiles(const filesystem::path &directory, bool recursive, bool directories) const
+{
+	filesystem::path relative = GetPathInZip(directory);
+	vector<filesystem::path> fileList;
+
+	lock_guard<recursive_mutex> guard(lock);
+	if(unzGoToFirstFile(zipFile) != UNZ_OK)
+		throw runtime_error("Failed to go to first file in ZIP");
+	do {
+		char filename[256];
+		unz_file_info fileInfo;
+		unzGetCurrentFileInfo(zipFile, &fileInfo, filename, sizeof(filename), NULL, 0, NULL, 0);
+
+		filesystem::path zipEntry = filename;
+		bool isDirectory = string(filename).back() == '/';
```

Comment:
Doesn't the conversion to string use the other slashes on Windows?
https://en.cppreference.com/w/cpp/filesystem/path/native
```suggestion
		bool isDirectory = filename.generic_string().back() == '/';
```
I'm guessing we can't use is_directory because we're inside the zip? If yes, maybe add a comment on it?

Reply:
We are inside of the zip file, that's not going to have windows slashes. Also, `filename` is a `char[]`, not an `std::filesystem::path` object.